### PR TITLE
clean up async methods and call chainings

### PR DESCRIPTION
**Pull request by maliroteh-sf**
We currently have many places in our repo where we are using anti async patterns:

- Functions that are async but not marked with `async` keyword
- Functions that are marked as async but are called in sync manner without calling `await()` on them
- Not taking advantage of Promise-chaining when having calls to multiple async function one after the other

This PR is to clean up all of these issues.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
                 "-n",
                 "c/helloWorld",
                 "-t",
-                "emu2"
```
Comment (maliroteh-sf) [inline .vscode/launch.json:31]: I think we should keep it consistent and use `sfdxdebug` everywhere in this file

File: src/common/AndroidLauncher.ts [hunk 2]
```diff
@@ -30,13 +30,8 @@ export class AndroidLauncher {
         const androidApi = preferredPack.platformAPI;
         const abi = preferredPack.abi;
         const device = androidConfig.supportedDevices[0];
-        let requestedPort = await AndroidSDKUtils.getNextAndroidAdbPort();
+        const requestedPort = await AndroidSDKUtils.getNextAndroidAdbPort();
         const spinner = cli.action;
-        // need to incr by 2, one for console port and next for adb
```
Comment (maliroteh-sf) [inline src/common/AndroidLauncher.ts:35]: move the incrementing logic into `getNextAndroidAdbPort()` b/c it is going to be replicated in multiple places.

File: src/common/CommonUtils.ts [hunk 3]
```diff
@@ -19,6 +20,20 @@ export class CommonUtils {
         return Promise.resolve();
     }
 
+    public static resolvePath(inputPath: string): string {
```
Comment (maliroteh-sf) [inline src/common/CommonUtils.ts:23]: created this b/c this logic of resolving a path is used in multiple places.

File: src/cli/commands/force/lightning/local/device/create.ts [hunk 4]
```diff
@@ -150,27 +150,30 @@ export class Create extends Setup {
...
-        const androidApi = preferredPack.platformAPI;
-        const abi = preferredPack.abi;
-
-        return AndroidSDKUtils.createNewVirtualDevice(
-            this.deviceName,
-            emuImage,
-            androidApi,
-            this.deviceType,
-            abi
+        return AndroidSDKUtils.findRequiredEmulatorImages().then(
```
Comment (khawkins) [inline src/cli/commands/force/lightning/local/device/create.ts:153]: For the record, I don't know that I have a preference *across the board* for this pattern over the other. I haven't yet developed a keen sense of when to use one over the other—some obvious cases (to me) are e.g. when there are async blocks nested 10 deep and as such increasingly hard to follow, or complex conditional async calls that might otherwise benefit from a "flattening". Other cases aren't so obvious to me.

To be clear, I think the change you've made here is fine. I just figured I'd throw out the follow-on thought processes I engaged around reviewing it.

File: src/cli/commands/force/lightning/local/device/create.ts [hunk 5]
```diff
@@ -150,27 +150,30 @@ export class Create extends Setup {
     }
 
     private async executeAndroidDeviceCreate(): Promise<any> {
```
Comment (khawkins) [inline src/cli/commands/force/lightning/local/device/create.ts:152]: Not sure if we'd do it as part of this exercise, but: should we make the return type `Promise<boolean>` here? That's what `AndroidSDKUtils.createNewVirtualDevice()` returns, which is what we're ultimately returning.

File: src/cli/commands/force/lightning/local/device/create.ts [hunk 6]
```diff
@@ -150,27 +150,30 @@ export class Create extends Setup {
...
+                    emuImage,
+                    androidApi,
+                    this.deviceType,
+                    abi
+                );
+            }
         );
     }
 
     private async executeIOSDeviceCreate(): Promise<any> {
```
Comment (khawkins) [inline src/cli/commands/force/lightning/local/device/create.ts:170]: Same question on return type here, except this would be `Promise<string>` in this case. That the two platforms return different types for the same conceptual action is another topic entirely. 😛

File: src/cli/commands/force/lightning/local/device/create.ts (hunk 4 above)
Comment (maliroteh-sf) [reply src/cli/commands/force/lightning/local/device/create.ts:153]: I'm glad you brought this up, b/c I thought of it too and then thought to myself that, if we ever have a function that is going to be that complex, then that would be our cue to break up the function into smaller functions. But yes, the idea of using chaining vs "flattened" approach should not be set in stone and we should evaluate on as needed basis.

File: src/cli/commands/force/lightning/local/device/list.ts [hunk 7]
```diff
@@ -62,29 +62,25 @@ export default class List extends SfdxCommand {
...
         }
 
-        return new Promise<any>((resolve, reject) => {
-            const deviceList = CommandLineUtils.platformFlagIsIOS(platform)
-                ? this.iOSDeviceList()
-                : this.androidDeviceList();
-
-            resolve(deviceList);
-        });
+        return CommandLineUtils.platformFlagIsIOS(platform)
```
Comment (khawkins) [inline src/cli/commands/force/lightning/local/device/list.ts:65]: This is another pattern that I'm uncertain about—more specifically, for the entirety of this method.  And it boils down to this: functionally/executionally, what's the difference between this:
```typescript
async function asyncThing(): Promise<boolean> {
    const retVal = false;
    return Promise.resolve(retVal);
}
```
And this:
```typescript
async function asyncThing(): Promise<boolean> {
    return new Promise<boolean>((resolve, reject) => {
        const retVal = false;
        resolve(retVal);
    });
}
```
Because my gut says that these will actually exhibit different behaviors—if subtly—at runtime. [This SO answer](https://stackoverflow.com/a/27716590) seems to run through the behavior differences? But I'm still parsing it to understand better what it means at runtime.

Comment (khawkins) [reply src/cli/commands/force/lightning/local/device/list.ts:65]: Oops, this is more applicable below. I'll move the comment to the right place.

File: src/cli/commands/force/lightning/local/device/list.ts [hunk 8]
```diff
@@ -62,29 +62,25 @@ export default class List extends SfdxCommand {
...
+            : this.androidDeviceList();
     }
 
     public async iOSDeviceList(): Promise<IOSSimulatorDevice[]> {
         performance.mark(this.perfMarker.startMarkName);
         const result = await IOSUtils.getSupportedSimulators();
         performance.mark(this.perfMarker.endMarkName);
         this.showDeviceList(result);
-        return result;
+        return Promise.resolve(result);
```
Comment (khawkins) [inline src/cli/commands/force/lightning/local/device/list.ts:75]: This is another pattern that I'm uncertain about—more specifically, for the entirety of this method.  And it boils down to this: functionally/executionally, what's the difference between this:
```typescript
async function asyncThing(): Promise<boolean> {
    const retVal = false;
    return Promise.resolve(retVal);
}
```
And this:
```typescript
async function asyncThing(): Promise<boolean> {
    return new Promise<boolean>((resolve, reject) => {
        const retVal = false;
        resolve(retVal);
    });
}
```
Because my gut says that these will actually exhibit different behaviors—if subtly—at runtime. [This SO answer](https://stackoverflow.com/a/27716590) seems to run through the behavior differences? But I'm still parsing it to understand better what it means at runtime.

File: src/common/AndroidUtils.ts [hunk 9]
```diff
@@ -74,78 +72,65 @@ export class AndroidSDKUtils {
...
     public static async androidSDKPrerequisitesCheck(): Promise<string> {
-        return new Promise(async (resolve, reject) => {
-            // Attempt to run sdkmanager and see if it throws any exceptions.
-            // If no errors are encountered then all prerequisites are met.
-            // But if an error is encountered then we'll try to see if it
-            // is due to unsupported Java version or something else.
-            AndroidSDKUtils.fetchAndroidCmdLineToolsLocation([
-                'ignore', // stdin
-                'pipe', // stdout
-                'pipe' // stderr
```
Comment (sfdctaka) [inline src/common/AndroidUtils.ts:85]: Just for my knowledge and want to understand. Why do we not need to pass `ignore, pipe, & pipe` anymore?

File: src/cli/commands/force/lightning/local/device/create.ts (hunk 4 above)
Comment (sfdctaka) [reply src/cli/commands/force/lightning/local/device/create.ts:153]: Inline `async` [here](https://github.com/forcedotcom/lwc-dev-mobile/blob/c14c1dc9539205405143f77e104373a1392c9e5b/src/common/AndroidUtils.ts#L187) I thought was unique.

About to finish reviewing AndroidUtil.ts. That's a long one 😌

File: src/cli/commands/force/lightning/local/device/list.ts (hunk 8 above)
Comment (sfdctaka) [reply src/cli/commands/force/lightning/local/device/list.ts:75]: How I view the two is that both are `then`-able but the second `asyncThing` example allows room to add any logic in the future. As for the real code in `IOSUtils.getSupportedSimulators()` it only returns a resolved value(even when it `catch`es; returns empty array) so simply returning that result in `Promise.resolve` is more compact than the second example, but een simpler, just return `result` as that is equivalent to `Promise.resolve`. I tested that on this branch and it works.

image

**Review by sfdctaka — COMMENTED** [review]
LGTM. As Kevin pointed out should we decide to take a further look and make changes for consistency on `Promise.resolve` I'm all for that.

File: src/cli/commands/force/lightning/local/device/list.ts (hunk 8 above)
Comment (khawkins) [reply src/cli/commands/force/lightning/local/device/list.ts:75]: That's my takeaway too, after more reading and writing of my own sample code to test the patterns out. 😀 This is probably another one that comes down to context, as to how we choose the code pattern.

Comment (khawkins) [reply src/cli/commands/force/lightning/local/device/list.ts:75]: To back up to the original line of code: I *definitely* support this change, of returning `Promise.resolve(result)` over simply returning `result`. Even if they end up functioning equivalently, I prefer the update to the more intentional code pattern.

File: src/common/AndroidUtils.ts (hunk 9 above)
Comment (khawkins) [reply src/common/AndroidUtils.ts:85]: 1. We switched `fetchAndroidCmdLineToolsLocation()` to ultimately call `child_process.exec()` (async), whereas it used to call `child_process.execSync()` (sync), and the async method does not support `StdioOptions`.
2. In this case, `ignore` for `stdin` never really mattered, and `pipe` for `stdout` and `stderr` is effectively the way that `CommonUtils.executeCommandAsync()` operates by default.

File: src/common/AndroidUtils.ts [hunk 10]
```diff
@@ -154,157 +139,147 @@ export class AndroidSDKUtils {
         }
 
         if (!AndroidSDKUtils.isCached()) {
```
Comment (khawkins) [inline src/common/AndroidUtils.ts:141]: Minor: with the switch to using more of an async pattern here, it might read better/less "indented" to switch to a quick return pattern instead:
```typescript
if (AndroidSDKUtils.isCached()) {
    return Promise.resolve(AndroidSDKUtils.packageCache);
}

return CommonUtils.executeCommandAsync(
// ...
```

File: src/common/AndroidUtils.ts [hunk 11]
```diff
@@ -154,157 +139,147 @@ export class AndroidSDKUtils {
...
+    public static async fetchEmulators(): Promise<AndroidVirtualDevice[]> {
+        let devices: AndroidVirtualDevice[] = [];
+        return CommonUtils.executeCommandAsync(
+            AndroidSDKUtils.getAvdManagerCommand() + ' list avd'
+        )
+            .then((result) => {
+                if (result.stdout && result.stdout.length > 0) {
                     devices = AndroidVirtualDevice.parseRawString(
-                        result.toString()
+                        result.stdout
```
Comment (khawkins) [inline src/common/AndroidUtils.ts:168]: Nice to learn that the default encoding of `child_process.exec()` is 'utf8', returning strings instead of `Buffer`s (the default of `execSync`). A lot less cumbersome to deal with in our use case.

File: src/common/AndroidUtils.ts [hunk 12]
```diff
@@ -154,157 +139,147 @@ export class AndroidSDKUtils {
...
                     );
                 }
-            } catch (exception) {
+                return Promise.resolve(devices);
+            })
+            .catch((exception) => {
```
Comment (khawkins) [inline src/common/AndroidUtils.ts:173]: Minor: as near as I can tell, this is the only place in our code base where we call the 'error' `exception`. May as well switch it to `error` like the rest of them. 🙂

Comment (khawkins) [reply src/common/AndroidUtils.ts:173]: (Okay I found a few other non-'error' examples, but most of the code uses 'error'. 😉)

File: src/cli/commands/force/lightning/local/device/list.ts (hunk 8 above)
Comment (khawkins) [reply src/cli/commands/force/lightning/local/device/list.ts:75]: Another thing I've learned about through this exploration: [the Deferred anti-pattern](https://github.com/petkaantonov/bluebird/wiki/Promise-anti-patterns#the-deferred-anti-pattern). Maybe I'll get good at this someday. 😉

File: src/common/AndroidUtils.ts [hunk 13]
```diff
@@ -154,157 +139,147 @@ export class AndroidSDKUtils {
...
-            const packages = await AndroidSDKUtils.fetchInstalledPackages();
-            if (packages.isEmpty()) {
-                return Promise.reject(
-                    new Error(
-                        `No Android API packages are installed. Minimum supported Android API package version is ${androidConfig.minSupportedRuntimeAndroid}`
-                    )
-                );
-            }
+        return AndroidSDKUtils.fetchInstalledPackages()
+            .then(async (packages) => {
```
Comment (khawkins) [inline src/common/AndroidUtils.ts:187]: What does `async` buy you here?

File: src/common/AndroidUtils.ts [hunk 14]
```diff
@@ -154,157 +139,147 @@ export class AndroidSDKUtils {
...
-                new Error(
-                    `Could not find android api packages. ${error.errorMessage}`
+                // If we got here then it means that we don't have any Android API packages with emulator images.
+                // So we will go ahead and return the latest one anyway. This is b/c the setup command will error
+                // out stating that no packages with matching emulator images have been found so the user would
+                // then know that they should install emulator images for the latest API package.
+                return Promise.resolve(matchingPlatforms[0]);
+            })
+            .catch((error) =>
+                Promise.reject(
```
Comment (khawkins) [inline src/common/AndroidUtils.ts:230]: Why would you not `return` the Promise here?

File: src/common/AndroidUtils.ts [hunk 15]
```diff
@@ -446,22 +418,17 @@ export class AndroidSDKUtils {
...
-        return new Promise((resolve, reject) => {
-            try {
-                CommonUtils.executeCommandSync(openUrlCommand);
-                resolve(true);
-            } catch (error) {
-                reject(error);
-            }
-        });
+        return CommonUtils.executeCommandAsync(openUrlCommand)
+            .then((result) => Promise.resolve(true))
```
Comment (khawkins) [inline src/common/AndroidUtils.ts:427]: You know, this isn't necessarily explicitly related to this PR, but sniffing this scenario of "return `true` but never return `false`", I followed this all the way back up to `Preview.launchPreview()`, called from `Preview.run()` (i.e. the top of the chain), and as near as I can tell there's no reason for this Promise chain to return a `boolean`. The Promise resolution value is never read, again as near as I can tell.  Food for thought.

File: src/common/AndroidUtils.ts (hunk 13 above)
Comment (maliroteh-sf) [reply src/common/AndroidUtils.ts:187]: on line 214 I'm using `await` and you can't do that without `async` here.

File: src/common/AndroidUtils.ts (hunk 14 above)
Comment (maliroteh-sf) [reply src/common/AndroidUtils.ts:230]: Hmm... looks like I need to work on the logic here a bit more. Previously we had a try-catch block b/c after going through all of the checks (which returned Promise.reject) we could still call into `AndroidSDKUtils.packageWithRequiredEmulatorImages()` which can throw instead of coming back with no results so for that scenario we were catching the thrown error and returning it here. But as you pointed out, I think we can use some more cleanup here.

File: src/common/AndroidUtils.ts (hunk 15 above)
Comment (maliroteh-sf) [reply src/common/AndroidUtils.ts:427]: yeah, I don't know why this was done this way either. I'm in favor of changing such cases to have `Promise<void>` return type. Will go ahead and make the changes.

File: src/common/AndroidUtils.ts [hunk 16]
```diff
@@ -74,237 +72,216 @@ export class AndroidSDKUtils {
...
-        return new Promise((resolve, reject) => {
-            let devices: AndroidVirtualDevice[] = [];
-            try {
-                const result = CommonUtils.executeCommandSync(
-                    AndroidSDKUtils.getAvdManagerCommand() + ' list avd'
-                );
-                if (result) {
+                return Promise.resolve(AndroidSDKUtils.packageCache);
+            })
+            .catch((error) => Promise.reject(error));
```
Comment (khawkins) [inline src/common/AndroidUtils.ts:157]: Not trying to boil the ocean here, but just throwing it out as more forward thinking food for thought around best practices: should we ditch `catch` blocks like this, that simply propagate the caught error? I don't believe they're necessary (except in `new Promise()` constructs. 😂)

File: src/common/AndroidUtils.ts [hunk 17]
```diff
@@ -288,7 +290,7 @@ export class AndroidSDKUtils {
         platformAPI: string,
         device: string,
         abi: string
-    ): Promise<boolean> {
+    ): Promise<void> {
```
Comment (sfdctaka) [inline src/common/AndroidUtils.ts:293]: Sharing my learning. I wanted to see if we can declare the type for `reject` but in Typescript it's apparently set to `any`. https://github.com/Microsoft/TypeScript/issues/7588#issuecomment-199079907
That's why it's fine that we pass`Error`. :)

File: src/common/AndroidUtils.ts (hunk 16 above)
Comment (sfdctaka) [reply src/common/AndroidUtils.ts:157]: ➕

File: src/common/AndroidUtils.ts [hunk 18]
```diff
@@ -362,106 +339,88 @@ export class AndroidSDKUtils {
...
         const openUrlCommand = `${AndroidSDKUtils.getAdbShellCommand()} -s emulator-${emulatorPort} shell am start -a android.intent.action.VIEW -d ${url}`;
-        return new Promise((resolve, reject) => {
-            try {
-                CommonUtils.executeCommandSync(openUrlCommand);
-                resolve(true);
-            } catch (error) {
-                reject(error);
-            }
-        });
+        return CommonUtils.executeCommandAsync(openUrlCommand)
```
Comment (khawkins) [inline src/common/AndroidUtils.ts:418]: This should just be `return CommonUtils.executeCommandAsync(openUrlCommand);`, right? The `then` and `catch` blocks below are just pass-through.

Comment (khawkins) [reply src/common/AndroidUtils.ts:418]: Oh never mind that's going to get into casting errors with the Promise signatures.

File: src/common/AndroidUtils.ts [hunk 19]
```diff
@@ -471,49 +430,52 @@ export class AndroidSDKUtils {
...
+            thePromise = CommonUtils.executeCommandAsync(installCommand);
+        } else {
+            thePromise = Promise.resolve({ stdout: '', stderr: '' });
+        }
 
-            let launchArgs =
-                `--es "${PreviewUtils.COMPONENT_NAME_ARG_PREFIX}" "${compName}"` +
-                ` --es "${PreviewUtils.PROJECT_DIR_ARG_PREFIX}" "${projectDir}"`;
+        return thePromise
+            .then(() => {
```
Comment (khawkins) [inline src/common/AndroidUtils.ts:447]: This feels weird. I know why it is the way it is, and I don't know of a better solution. But it still feels weird.

File: src/common/CommonUtils.ts [hunk 20]
```diff
@@ -19,6 +20,24 @@ export class CommonUtils {
         return Promise.resolve();
     }
 
+    public static delay(ms: number): Promise<void> {
+        return new Promise((resolve) => setTimeout(resolve, ms));
+    }
+
+    public static resolvePath(inputPath: string): string {
```
Comment (khawkins) [inline src/common/CommonUtils.ts:27]: Can we rename this to `resolveUserHomePath`?

File: src/common/CommonUtils.ts [hunk 21]
```diff
@@ -19,6 +20,24 @@ export class CommonUtils {
...
+        let newPath = inputPath.trim();
+        if (newPath.startsWith('~')) {
+            const USER_HOME =
```
Comment (khawkins) [inline src/common/CommonUtils.ts:30]: Let's switch to a more localized nomenclature: `userHome`.

File: src/common/CommonUtils.ts (hunk 20 above)
Comment (maliroteh-sf) [reply src/common/CommonUtils.ts:27]: well it's not just resolving the home path, that's just one of the things it does.
1. Checks for and resolves the home path
2. Then checks to resolve the path (i.e. if it is relative then it is resolved to an absolute path)
3. Then does a normalization of the final path before returning it.

I couldn't come up with a better name for this function though 😄

File: src/common/AndroidUtils.ts (hunk 19 above)
Comment (maliroteh-sf) [reply src/common/AndroidUtils.ts:447]: yeah, I tried to come up with something better but couldn't think of something.

File: src/common/CommonUtils.ts (hunk 20 above)
Comment (khawkins) [inline src/common/CommonUtils.ts:27]: I'm also a little dubious of the extent of the functionality in here. We're only calling this—and likely to only call this—in scenarios where we know we need to resolve the various locations that could apply to the user's home directory.

IMO we should just retool this function to return the user's home directory. The rest of it seems like a solution in search of a problem.

File: src/common/AndroidUtils.ts (hunk 18 above)
Comment (maliroteh-sf) [reply src/common/AndroidUtils.ts:418]: Not exactly. This functions returns `Promise<void>` but `executeCommandAsync` doesn't. So I need to keep the `then` but the `catch` can go.

File: src/common/AndroidUtils.ts (hunk 16 above)
Comment (maliroteh-sf) [reply src/common/AndroidUtils.ts:157]: Yes you're right. I'll clean these up.

File: src/common/CommonUtils.ts [hunk 22]
```diff
@@ -49,7 +68,16 @@ export class CommonUtils {
                             `Error executing command '${command}':`
                         );
                         CommonUtils.logger.error(`${error}`);
-                        reject(error);
+
+                        // also include stderr & stdout for more detailed error
+                        let msg = error.message;
+                        if (stderr && stderr.length > 0) {
+                            msg = `${msg}\n${stderr}`;
```
Comment (khawkins) [inline src/common/CommonUtils.ts:75]: You probably want to identify these in the message too, e.g. ``msg = `${msg}\nstderr:\n${stderr}`;``, etc. I'm afraid the error message could become something of a jumbled mess otherwise.

Comment (khawkins) [reply src/common/CommonUtils.ts:75]: Also, though, I'm not sure that this is set up the way we want it. It feels like we should maybe log the additional information, but simply return the original error as output. This is a pretty general command, and depending on how consumers use it, it could end up dumping a lot of information out to e.g. a console or other UI, in a way that consumers couldn't really control.

File: src/common/CommonUtils.ts (hunk 20 above)
Comment (maliroteh-sf) [reply src/common/CommonUtils.ts:27]: In this repo yes. But in the internal repo we're going to be needing to do the 3 logic here for example for the configure certificates command. I thought it would be useful to have the logic in a function in common utilities so that after breaking up this repo, any repo that's going to need this logic can just consume it from CommonUtils.

What do you think. Should I still go ahead and revert this?

Comment (khawkins) [reply src/common/CommonUtils.ts:27]: I am very likely missing something in the details, but it feels like—at least from the way we're currently consuming it—that we're duplicating/replicating efforts that the consumer is already making.

For example, the typical input that this method is currently receiving is along the lines of `~/.android/avd/myEmu.avd/emu-launch-params.txt`. It seems like the *only* thing we don't know from that input is what `~` is supposed to resolve to.

Now if we envision future scenarios where we're going to pass less literal, more extensible paths into this method, like say:
```typescript
// anotherDerivedPath got set to something like '../Projects/deep/down/../top-config.txt',
// out of our purview.
const fullPath = CommonUtils.resolvePath(path.join('~', anotherDerivedPath))
```
Then it makes sense to me to have the additional functionality.  As it's currently being used, its primary function seems to be dereferencing the proper home path.

Comment (maliroteh-sf) [reply src/common/CommonUtils.ts:27]: You know what, you've convinced me. Let's just simplify this method and make it to only resolve the user home path. Anything else can be done by chaining `path.resolve` and `path.normalize` by the caller.

File: src/common/CommonUtils.ts (hunk 22 above)
Comment (maliroteh-sf) [reply src/common/CommonUtils.ts:75]: Very good point. I totally didn't think about that. Ok will update the logic here to only return the `error` itself but additionally log the `error + stderr + stdout`

File: src/common/AndroidLauncher.ts [hunk 23]
```diff
@@ -106,13 +106,3 @@ export class AndroidLauncher {
             });
     }
 }
-
-// let launcher = new AndroidLauncher('testemu7');
```
Comment (khawkins) [inline src/common/AndroidLauncher.ts:115]: End of an era! 😉

**Review by khawkins — APPROVED** [review]
💯 Thanks for all of your hard work on this one! That was quite the exercise, but I think the code is a lot better off after doing it!
